cmd/juju/charmhub/download.go

```diff
@@ -286,7 +252,60 @@ Install the %q %s with:
 	return nil
 }
 
-func (c *downloadCommand) suggested(requestedSeries string, channel string, releases []transport.Release, cmdContext *cmd.Context) error {
+func (c *downloadCommand) refresh(ctx context.Context, cmdContext *cmd.Context, client CharmHubClient, normChannel charm.Channel, arch, series string, retrySuggested bool) ([]transport.RefreshResponse, *corecharm.Platform, error) {
+	base, err := coreseries.GetBaseFromSeries(series)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	platform := fmt.Sprintf("%s/%s/%s", arch, base.Name, base.Channel.Track)
+	normBase, err := corecharm.ParsePlatformNormalize(platform)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	// Ensure we compute the base channel correctly.
+	computedNormBase := corecharm.ComputeBaseChannel(normBase)
+
+	refreshConfig, err := charmhub.InstallOneFromChannel(c.charmOrBundle, normChannel.String(), charmhub.RefreshBase{
+		Architecture: computedNormBase.Architecture,
+		Name:         computedNormBase.OS,
+		Channel:      computedNormBase.Channel,
+	})
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	results, err := client.Refresh(ctx, refreshConfig)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	if len(results) == 0 {
+		return nil, nil, errors.NotFoundf(c.charmOrBundle)
+	}
+	// Ensure we didn't get any errors whilst querying the charmhub API
+	for _, res := range results {
+		if res.Error != nil {
+			if res.Error.Code == transport.ErrorCodeRevisionNotFound {
+				possibleSeries, err := c.suggested(cmdContext, series, normChannel.String(), res.Error.Extra.Releases)
+				// The following will attempt to refresh the charm with the
+				// suggested series. If it can't do that, it will give up after
+				// the second attempt.
+				if retrySuggested && errors.Is(err, errors.NotSupported) && len(possibleSeries) > 0 {
+					cmdContext.Infof("Series %q is not supported for charm %q, trying series %q", series, c.charmOrBundle, possibleSeries[0])
+					return c.refresh(ctx, cmdContext, client, normChannel, arch, possibleSeries[0], false)
```

Comment:
Using possibleSeries[0] can lead to an older charm. Suggest the most recently series instead.

Reply:
I think we should just use what the api tells us. In the case of the ghost charm it sent back xenial then trusty, which is what I would expect. Also how do order centos over Ubuntu charms?